### PR TITLE
HBASE-22393 HBOSS: Shaded external dependencies to avoid conflicts with Hadoop and HBase.

### PR DESCRIPTION
**Pull request by busbey**
Compared to the patch "0001-HBASE-22393.patch" on [HBASE-22393](https://issues.apache.org/jira/browse/HBASE-22393)

* Correct commit message
* Corrected Author / Co-authored-by infromation
* Included missed top level pom changes from "0001-HBASE-22393-HBOSS-Shade-dependencies-to-prevent-conf.patch" patch
* Corrected LICENSE/NOTICE on generated shaded jar
* Moved log4j specific dependencies to test scope
* Corrected leaving slf4j-api as a dependency post-shading
* Corrected shading for audience-annotations and netty
* Corrected excluding some unneeded transitive dependencies
* Reordered some dependencies to improve readability for dependency plugin output
* Added a parent pom at the top level to clear warnings about plugin versions

### CODE REVIEW

**Comment by wchevreuil** [comment]
Ran maven tests for both hadoop2 and hadoop3 profiles, both working.
Deployed built jar on a test cluster: 
- Cluster restart went ok, no previously NoClassDefFoundError due to classpath issues seen;
- hbck execution is fine, no previously NoClassDefFoundError due to classpath issues seen;
- Executed some ltt runs, no issues seen.

lgtm, +1

File: hbase-oss/pom.xml [hunk 1]
```diff
@@ -70,56 +71,206 @@
...
+                  <shadedPattern>${shading.prefix}.org.codehaus</shadedPattern>
+                </relocation>
+                <!-- netty family -->
+                <relocation>
+                    <pattern>org.jboss.netty</pattern>
+                    <shadedPattern>${shading.prefix}.org.jboss.netty</shadedPattern>
+                </relocation>
+                <relocation>
+                    <pattern>io.netty</pattern>
+                    <shadedPattern>${shading.prefix}.io.netty</shadedPattern>
```
Comment (joshelser) [inline hbase-oss/pom.xml]: Do we need to advise users about `io.netty.packagePrefix` and rename any of the shared objects from Netty?

I'm reminded of what we have in hbase-thirdparty: https://github.com/apache/hbase-thirdparty/blob/master/hbase-shaded-netty/pom.xml#L93-L137

File: hbase-oss/src/main/appended-resources/META-INF/NOTICE [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+    * http://www.jcraft.com/jzlib/
+
+This product contains a modified version of 'Webbit', a Java event based
+WebSocket and HTTP server, under a BSD license.
+
+  * HOMEPAGE:
+    * https://github.com/joewalnes/webbit
+
+----
+This work bundles additional works licensed under the Apache License v2.0.
```
Comment (joshelser) [inline hbase-oss/src/main/appended-resources/META-INF/NOTICE]: I remember seeing some LEGAL jira issue which said that copyright only isn't required to propogate in NOTICE, but it isn't hurting anything.

File: hbase-oss/pom.xml [hunk 3]
```diff
@@ -70,56 +71,206 @@
...
+        <artifactId>maven-shade-plugin</artifactId>
+        <executions>
+          <execution>
+            <id>aggregate-into-a-jar-with-relocated-third-parties</id>
+            <phase>package</phase>
+            <goals>
+              <goal>shade</goal>
+            </goals>
+            <configuration>
+              <createDependencyReducedPom>true</createDependencyReducedPom>
```
Comment (joshelser) [inline hbase-oss/pom.xml]: Please also configure `dependencyReducedPomLocation` as per https://maven.apache.org/plugins/maven-shade-plugin/shade-mojo.html#dependencyReducedPomLocation to move it inside of `${project.build.dir}` (more commonly known as:`target/`)

Comment (busbey) [reply hbase-oss/pom.xml]: setting the location doesn't actually work. it's a long standing problem with the shade plugin that keeps getting only half fixed.

this is why we don't change it's location in the main project repo's shaded jar work. (or over in the hadoop project either)

File: hbase-oss/src/main/appended-resources/META-INF/NOTICE (hunk 2 above)
Comment (busbey) [reply hbase-oss/src/main/appended-resources/META-INF/NOTICE]: I remember the discussion you're talking about. It was ASF copyrights specifically. I added the ZK one here to keep it consistent with the automated NOTICE roll-up from our other ASF dependencies.

We could instead manually filter ASF project sourced NOTICE files and maintain a copyright notice at the top that covered the full range of dates, but that all seems like more work than it's worth.

File: hbase-oss/pom.xml (hunk 3 above)
Comment (joshelser) [reply hbase-oss/pom.xml]: Really? I wonder how I've thought this was working elsewhere then.. Strange.

What about copying the clean-plugin configuration to get rid of it and updating .gitignore then? (as a follow on)

File: hbase-oss/pom.xml (hunk 1 above)
Comment (busbey) [reply hbase-oss/pom.xml]: IIRC that's a Netty 4 issue and not present in Netty 3. The jar here has no JNI shared objects, fwiw.

File: hbase-oss/src/main/appended-resources/META-INF/NOTICE (hunk 2 above)
Comment (joshelser) [reply hbase-oss/src/main/appended-resources/META-INF/NOTICE]: > that all seems like more work than it's worth.
True that.

No need to undo anything auto-magic. Fine the way it is now.

File: hbase-oss/pom.xml (hunk 1 above)
Comment (joshelser) [reply hbase-oss/pom.xml]: Ah! I thought `io.netty` was only introduced in Netty4, but I see I was mistaken. Agree with your assessment.

**Pull request closed without merging**